op-service/txmgr/txmgr.go

```diff
@@ -212,52 +225,97 @@ func (m *SimpleTxManager) send(ctx context.Context, candidate TxCandidate) (*typ
 func (m *SimpleTxManager) craftTx(ctx context.Context, candidate TxCandidate) (*types.Transaction, error) {
-	gasTipCap, basefee, err := m.suggestGasPriceCaps(ctx)
+	gasTipCap, basefee, blobBasefee, err := m.suggestGasPriceCaps(ctx)
 	if err != nil {
 		m.metr.RPCError()
 		return nil, fmt.Errorf("failed to get gas price info: %w", err)
 	}
 	gasFeeCap := calcGasFeeCap(basefee, gasTipCap)
 
-	rawTx := &types.DynamicFeeTx{
-		ChainID:   m.chainID,
-		To:        candidate.To,
-		GasTipCap: gasTipCap,
-		GasFeeCap: gasFeeCap,
-		Data:      candidate.TxData,
-		Value:     candidate.Value,
-	}
-
-	m.l.Info("Creating tx", "to", rawTx.To, "from", m.cfg.From)
+	gasLimit := candidate.GasLimit
 
 	// If the gas limit is set, we can use that as the gas
-	if candidate.GasLimit != 0 {
-		rawTx.Gas = candidate.GasLimit
-	} else {
+	if gasLimit == 0 {
 		// Calculate the intrinsic gas for the transaction
 		gas, err := m.backend.EstimateGas(ctx, ethereum.CallMsg{
 			From:      m.cfg.From,
 			To:        candidate.To,
 			GasTipCap: gasTipCap,
 			GasFeeCap: gasFeeCap,
-			Data:      rawTx.Data,
-			Value:     rawTx.Value,
+			Data:      candidate.TxData,
+			Value:     candidate.Value,
 		})
 		if err != nil {
 			return nil, fmt.Errorf("failed to estimate gas: %w", err)
 		}
-		rawTx.Gas = gas
+		gasLimit = gas
+	}
+
+	var sidecar *types.BlobTxSidecar
+	var blobHashes []common.Hash
+	if len(candidate.Blobs) > 0 {
+		if candidate.To == nil {
+			return nil, errors.New("blob txs cannot deploy contracts")
+		}
+		blobHashes = make([]common.Hash, 0, len(candidate.Blobs))
+		sidecar = &types.BlobTxSidecar{}
+		for i, blob := range candidate.Blobs {
+			rawBlob := *blob.KZGBlob()
+			sidecar.Blobs = append(sidecar.Blobs, rawBlob)
+			commitment, err := kzg4844.BlobToCommitment(rawBlob)
+			if err != nil {
+				return nil, fmt.Errorf("cannot compute KZG commitment of blob %d in tx candidate: %w", i, err)
+			}
+			sidecar.Commitments = append(sidecar.Commitments, commitment)
+			proof, err := kzg4844.ComputeBlobProof(rawBlob, commitment)
+			if err != nil {
+				return nil, fmt.Errorf("cannot compute KZG proof for fast commitment verification of blob %d in tx candidate: %w", i, err)
+			}
+			sidecar.Proofs = append(sidecar.Proofs, proof)
+			blobHashes = append(blobHashes, eth.KZGToVersionedHash(commitment))
+		}
 	}
 
-	return m.signWithNextNonce(ctx, rawTx)
+	var txMessage types.TxData
+	if sidecar != nil {
+		if blobBasefee == nil {
+			return nil, fmt.Errorf("expected non-nil blobBasefee")
+		}
+		blobFeeCap := calcBlobFeeCap(blobBasefee)
+		message := &types.BlobTx{
+			Nonce:      0, // set by signer
+			To:         *candidate.To,
+			Data:       candidate.TxData,
+			Gas:        gasLimit,
+			BlobHashes: blobHashes,
+			Sidecar:    sidecar,
+		}
+		if err := finishBlobTx(message, m.chainID, gasTipCap, gasFeeCap, blobFeeCap, candidate.Value); err != nil {
+			return nil, fmt.Errorf("%w: failed to create blob transaction", err)
```

Comment:
I think
```suggestion
			return nil, fmt.Errorf("failed to create blob transaction: %w", err)
```
would be more standard, so that the overflowing field is shown at the end of the error chain message.

Reply:
I've seen both patterns and have been consistently trying to use "%w:" as a prefix rather than suffix, but no opinion on which one to prefer and happy to adopt whichever is most idiomatic for op-stack (or in general).

Reply:
Thanks! Yes we usually suffix wrapped errors, which is also how golang itself is usually chaining error messages in the stdlib. This way, the error message consistently goes into more detail from left to right.